SignalServiceKit/src/Util/OWSAnalytics.m

```diff
@@ -145,46 +162,103 @@ - (void)tryToSyncEvents
+            eventKey:eventKey
+            success:^{
+                DDLogDebug(@"%@ sendEvent[critical] succeeded: %@", self.tag, eventKey);
+            }
+            failure:^{
+                DDLogError(@"%@ sendEvent[critical] failed: %@", self.tag, eventKey);
+            }];
+    } else {
         self.hasRequestInFlight = YES;
+        [self submitEvent:eventDictionary
+            eventKey:eventKey
+            success:^{
+                DDLogDebug(@"%@ sendEvent succeeded: %@", self.tag, eventKey);
+                dispatch_async(self.serialQueue, ^{
+                    self.hasRequestInFlight = NO;
+
+                    [self.dbConnection readWriteWithBlock:^(YapDatabaseReadWriteTransaction *transaction) {
+                        // Remove from queue.
+                        [transaction removeObjectForKey:eventKey inCollection:kOWSAnalytics_EventsCollection];
+                    }];
+
+                    // Wait a second between network requests / retries.
+                    dispatch_after(
+                        dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1.f * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
+                            [self tryToSyncEvents];
+                        });
+                });
+            }
+            failure:^{
+                DDLogError(@"%@ sendEvent failed: %@", self.tag, eventKey);
+                dispatch_async(self.serialQueue, ^{
+                    self.hasRequestInFlight = NO;
+
+                    // Wait a second between network requests / retries.
+                    dispatch_after(
+                        dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1.f * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
+                            [self tryToSyncEvents];
+                        });
+                });
+            }];
+    }
+}
 
-        __block UIBackgroundTaskIdentifier task;
-        task = [UIApplication.sharedApplication beginBackgroundTaskWithExpirationHandler:^{
-            self.hasRequestInFlight = NO;
+- (void)submitEvent:(NSDictionary *)eventDictionary
+           eventKey:(NSString *)eventKey
+            success:(void (^_Nonnull)())successBlock
+            failure:(void (^_Nonnull)())failureBlock
+{
+    OWSAssert(eventDictionary);
+    OWSAssert(eventKey);
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
+    OWSAssert(dispatch_get_current_queue() == self.serialQueue);
+#pragma clang diagnostic pop
 
-            [UIApplication.sharedApplication endBackgroundTask:task];
-        }];
+    DDLogDebug(@"%@ submitting: %@", self.tag, eventKey);
 
-        // Until we integrate with an analytics platform, behave as though all event delivery succeeds.
-        dispatch_async(dispatch_get_main_queue(), ^{
-            self.hasRequestInFlight = NO;
+    __block UIBackgroundTaskIdentifier task;
+    task = [UIApplication.sharedApplication beginBackgroundTaskWithExpirationHandler:^{
+        failureBlock();
```

Comment:
Does this run the risk of calling multiple completion handlers? e.g. either failure twice or failure once and success once?

Should we inspect/set a `hasCompleted` boolean before calling any of the failure/success blocks?